File: banchus-common/src/main/java/pe/nanamochi/banchus/domain/dto/OsuDirectQuery.java
package pe.nanamochi.banchus.domain.dto;

import java.util.HashMap;
import java.util.Map;
import lombok.AllArgsConstructor;
import lombok.Getter;

@Getter
@AllArgsConstructor
public enum OsuDirectQuery {
  NEWEST("Newest", "ranked_date:desc"),
  TOP_RATED("Top Rated", "favourite_count:desc"),
  MOST_PLAYED("Most Played", "play_count:desc");

  private final String query;
  private final String sort;

  private static final Map<String, OsuDirectQuery> LOOKUP = new HashMap<>();

  static {
    for (OsuDirectQuery q : values()) {
      LOOKUP.put(q.query.toLowerCase(), q);
    }
  }

  public static OsuDirectQuery fromQuery(String query) {
    if (query == null) return null;
    return LOOKUP.get(query.trim().toLowerCase());
  }
}

File: banchus-app/src/main/kotlin/pe/nanamochi/banchus/beatmap/enums/OsuDirectQuery.kt
package pe.nanamochi.banchus.beatmap.enums

enum class OsuDirectQuery(val query: String, val sort: String) {
    NEWEST("Newest", "ranked_date:desc"),
    TOP_RATED("Top Rated", "favourite_count:desc"),
    MOST_PLAYED("Most Played", "play_count:desc");

    companion object {
        fun fromQuery(query: String?): OsuDirectQuery? {
            if (query == null) return null
            val normalizedQuery = query.trim().lowercase()
            return entries.find { it.query.lowercase() == normalizedQuery }
        }
    }
}
